File: src/lib/software/obs_studio.py
from re import compile
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class ObsStudio(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'obs-studio'
        self.name = 'OBS Studio'
        self.category = SoftwareCategory.Media
        self.download_name = 'OBS-Studio-Windows-Installer.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'obs-studio.png'
        self.homepage = 'https://obsproject.com'

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        download_url_pattern = compile(r'https://cdn-fastly\.obsproject\.com/downloads/OBS-Studio-\d+.\d+.\d+-Windows-Installer\.exe')
        html = reply.readAll().data().decode()
        match = download_url_pattern.search(html)
        if not match:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
        else:
            self.url_resolved.emit(match.group(0))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest('https://obsproject.com/download'))

File: packages/app/src/services/software/definitions/obsStudio.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'obs-studio';
	public name = 'OBS Studio';
	public category = [SoftwareCategory.Media];
	public downloadName = 'OBS-Studio-Windows-Installer.exe';
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public icon = 'obs-studio.png';
	public homepage = 'https://obsproject.com';

	public async resolveDownloadUrl() {
		const res = await fetch('https://obsproject.com/download');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/cdn-fastly\.obsproject\.com\/downloads\/OBS-Studio-\d+.\d+.\d+-Windows-Installer\.exe/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[0]);
	}
}
